### ISSUE TITLE
Wrong detection: Python project detected as NodeJS

### ISSUE THREAD

**Issue by kadel** [redhat-developer/alizer#147]
https://github.com/blues-man/pipelines-vote-ui is detected as NodeJS application instead of Python

using Go Alizer ad2d3056c08d

```
# Runing recognizer.Analyzer("pipelines-vote-ui")
  {Name:JavaScript Aliases:[js node nodejs TypeScript] Weight:66.66 Frameworks:[] Tools:[] CanBeComponent:true}
  {Name:Python Aliases:[python3 rusthon] Weight:33.33 Frameworks:[] Tools:[] CanBeComponent:true}

# Runing recognizer.SelectDevFileFromTypes("pipelines-vote-ui")
  {Name:nodejs Language:JavaScript ProjectType:Node.js Tags:[Node.js Express ubi8]}
```

**Update @thepetk** :
Target date: 14 Apr 2023

**Comment by mohitsuman** [comment]
@elsony This issue impacts odo and OpenShift toolkit. Can we prioritise this and release an update.

**Comment by thepetk** [comment]
## TL;DR
The issue for this specific project is that we don't have implemented yet a `FlaskDetector`.

As the project has an extensive usage of `Javascript` inside its `templates/` folder, `alizer` recognizes `javascript` (weight: 66%) and `python` (weight: 33%).

So the `recognizer.SelectDevFileFromTypes` func selects the `nodeJS` stack instead of the `python` one. If `flask` was recognized as framework, the `python` stack would have been selected.

## Details
The devfile detection process:
1) The `SelectDevFileFromTypes` calls the `SelectDevFilesFromTypes` in order to have an array of related devfiles.

2) The [SelectDevFilesFromTypes](https://github.com/redhat-developer/alizer/blob/ad2d3056c08d511e132c67ed8e353a8fc67d5927/go/pkg/apis/recognizer/devfile_recognizer.go#L26) tries to analyze all components and get the correct `devfile` from them. In our case it finds **no component**. 

```go
        devFilesIndexes := selectDevFilesFromComponentsDetectedInPath(path, devFileTypes)
        if len(devFilesIndexes) > 0 {
		return devFilesIndexes, nil
        }
```

3) Finally, _(if no component was found)_ it tries to analyze all languages and get the correct `devfile` from this analysis. In our case it selects the `nodeJS` one as this is the language with the biggest weight and first in the list.
```go
	languages, err := analyze(path, &ctx)
	if err != nil {
		return []int{}, err
	}
	devfile, err := SelectDevFileUsingLanguagesFromTypes(languages, devFileTypes)
	if err != nil {
		return []int{}, errors.New("No valid devfile found for project in " + path)
	}
	return []int{devfile}, nil
```
As a result if the `flask` component was recognized from the `alizer` we would have matched the correct devfile.

## Further thinking
Although I think it is mandatory to add the `flask_detector` for now, another workaround for the future here, would be to check for the correct `devfile` through the `Dockerfile`. It may requires more effort but IMO it would enhance the functionality of `devfile` selection. Moreover, after the `selectDevFilesFromComponentsDetectedInPath` check we could check for any existing `dockerfile` in the root folder (or more in depth) and check there if we can get more info about the `devfile` selection.

**Comment by kadel** [comment]
> As `Python` has no list of `configuration_files` inside [languages-customization.yaml](https://github.com/redhat-developer/alizer/blob/main/go/pkg/utils/langfiles/resources/languages-customization.yml) we are not able to recognize any component.

I think that we could put `requirements.txt` and `pyproject.toml` as configuration_files for python. It will not cover 100% of python projects, but they are probably the two most common files in python projects. 



>As a result, without any components recognized, the next step is to select a devfile based on the language analysis. The project has an extensive usage of Javascript inside its templates/ directory, so alizer recognizes javascript (weight: 66%) and python (weight: 33%).

This is quite a common problem. I've seen a lot of projects incorrectly detected due to a number of HTML or JS files. I was wondering if it would make sense to give files in the root of the project more weight compared to files in subdirectories. It will require a lot of testing and tuning, but it could improve the detection.

**Comment by thepetk** [comment]
I have a draft [PR here](https://github.com/redhat-developer/alizer/pull/156) and I've tried several things in order to find the optimal solution.

> I think that we could put requirements.txt and pyproject.toml as configuration_files for python. It will not cover 100% of python projects, but they are probably the two most common files in python projects.
I've tried to put `manage.py`, `app.py` and `wsgi.py` (covers Django, Flask cases) and it worked for this case. I think with requirements might be better.

> I was wondering if it would make sense to give files in the root of the project more weight compared to files in subdirectories
Exactly. Right now we count the files found with the language's extension and we calculate the weight (for example in this case we have 2 `.js` files and 1 `.py`).

I have some doubts here, because updating the weights calculation method will have an overall impact, so as a result it will require more and more testing.

I was thinking maybe is better to give less weight to `.js` found inside a `templates/` directory.

**Comment by kadel** [comment]
> I've tried to put `manage.py`, `app.py` and `wsgi.py` (covers Django, Flask cases) and it worked for this case.

Probably just a nitpick, but, the field is called `configuration_files` putting source code files looks weird to me. If there are other files than configuration files, it might make sense to rename the field to something like `signal_files`. 

> I was thinking maybe is better to give less weight to .js found inside a templates/ directory.

That makes sense. It will just need to be extended to more directories than just `templates`. For example some projects use `static`, and the list might grew in future

**Comment by thepetk** [comment]
> Probably just a nitpick, but, the field is called configuration_files putting source code files looks weird to me. If there are other files than configuration files, it might make sense to rename the field to something like signal_files.

Normally, for python projects I would call a `configuration_file` the `app/settings.py` for `django` and the `config/config.py` for `flask`. Maybe I'm misinterpreting the term.

But in any case I think is good to have a workaround for `requirements.txt` and `project.toml`.

> That makes sense. It will just need to be extended to more directories than just templates. For example some projects use static, and the list might grew in future.

Yep `static` too (especially for django).

**Comment by thepetk** [comment]
I've finalized my [PR here](https://github.com/redhat-developer/alizer/pull/156). I have added this specific case into the e2e tests.

**Comment by thepetk** [comment]
I have also created a [dedicated issue](https://github.com/redhat-developer/alizer/issues/157) for the `templates`/`static` directory issue

**Comment by thepetk** [comment]
UPDATE: After the PR [comment](https://github.com/redhat-developer/alizer/pull/156#issuecomment-1491862605) I've created a new issue (https://github.com/redhat-developer/alizer/issues/165) which separates the blocking part.

On the other hand, I've closed (https://github.com/redhat-developer/alizer/issues/157) in order to continue working here for the language analysis part.

**Comment by thepetk** [comment]
FYI I've opened a [PR](https://github.com/redhat-developer/alizer/pull/168) for this issue. It replaces the logic of extension file count by giving points to each extension, while checking if they are part of a `templates/` or `static/` folder.
